### ISSUE TITLE
Getting error running filter_pos.four_criteria.py

### ISSUE THREAD

**Issue by jarcasariego** [lyijin/working_with_dna_meth#1]
Hello Dr. Liew, 

I'm attempting to characterize methylation in the coral M. cavernosa and giving a try to your pipeline. So far I successfully ran the code until the filtering with the four criteria step. The error is as follows:

 File "./filter_pos.four_criteria.py", line 46, in <module>
    scaf_size = int(scaf.split('size')[-1])
ValueError: invalid literal for int() with base 10: 'Sc0000000'

My understanding is that it is trying to extract the size of the scaffold from the first column. However, I cannot see the size in the bismark output (although it was in the header of the sequences in the genome fasta). 
I'm not very skilled with Python and maybe is a simple fix. Could you help me with this? 

Best, 

Javier

**Comment by lyijin** [comment]
Dear Javier,

Yeah, things do go a bit awry when my script is ran on genomes that doesn't fit a certain naming pattern.

On the genomes I've worked on, the scaffold names are named along the lines of "scaffold12345|size987654". what line 46 is trying to do is splitting the string by the word "size", then taking the last member of the array (that's what "-1" means in Python), then converting it to an integer. i.e.,

"scaffold12345|size987654" --> ['scaffold12345|', '987654'] --> '987654' --> 987654

When your scaffold has no "size" in it, the code will work till a point, and it fails because you can't convert a string into an integer.

"Sc0000000" --> ['Sc0000000'] --> 'Sc0000000' --> crash.

If RAM is plentiful on your system, and you know the maximum size of your scaffolds (let's say it's 123123123), simply do

(line 46)
`scaf_size = 123123123`

(line 50)
`bona_fide_pos[scaf] = np.zeros(scaf_size, dtype=np.int8)`

Let me know whether this works out, or if you get further crashes downstream.

**Comment by jarcasariego** [comment]
That worked perfectly. I was able to complete the filtration and confirm the selected loci with SeqMonk

As you anticipated the annotation code gave errors. I checked the issue and it seems that in the gff3 file there are several gene entries that have a wring end_position. This seems to be caused by including some exons from another scaffold. 

this is one of the over 300 examples: 

''jkavirodr@JARC-Surface:/mnt/c/Users/User/Downloads$ grep "Mcavernosa14088" Mcavernosa.maker.coding.gff3
Sc0000170       maker   gene    534277  25247   .       +       .       ID=Mcavernosa14088;Name=Mcavernosa14088;Alias=snap_masked-NewChr13-processed-gene-94.6;
Sc0000170       maker   mRNA    534277  25247   .       +       .       ID=Mcavernosa14088-RA;Parent=Mcavernosa14088;Name=Mcavernosa14088-RA;Alias=snap_masked-NewChr13-processed-gene-94.6-mRNA-1;_AED=0.73;_QI=0|0|0|0.33|1|1|6|0|205;_eAED=0.86;
Sc0000170       maker   exon    534277  534414  .       +       .       ID=Mcavernosa14088-RA:exon:4654;Parent=Mcavernosa14088-RA;
Sc0000170       maker   exon    535843  535950  .       +       .       ID=Mcavernosa14088-RA:exon:4655;Parent=Mcavernosa14088-RA;
Sc0000170       maker   exon    536292  536312  .       +       .       ID=Mcavernosa14088-RA:exon:4656;Parent=Mcavernosa14088-RA;
Sc0000171       maker   exon    20794   20832   .       +       .       ID=Mcavernosa14088-RA:exon:4657;Parent=Mcavernosa14088-RA;
Sc0000171       maker   exon    22025   22047   .       +       .       ID=Mcavernosa14088-RA:exon:4658;Parent=Mcavernosa14088-RA;
Sc0000171       maker   exon    24959   25247   .       +       .       ID=Mcavernosa14088-RA:exon:4659;Parent=Mcavernosa14088-RA;
Sc0000170       maker   CDS     534277  534414  .       +       0       ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
Sc0000170       maker   CDS     535843  535950  .       +       0       ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
Sc0000170       maker   CDS     536292  536312  .       +       0       ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
Sc0000171       maker   CDS     20794   20832   .       +       0       ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
Sc0000171       maker   CDS     22025   22047   .       +       0       ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
Sc0000171       maker   CDS     24959   25247   .       +       1       ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;''

I know this has nothing to do with your code but have you ever seem something like this? Do you have an idea to fix it? Should I fix it or simply ignore these genes? 

Thanks so much for your help.

**Comment by lyijin** [comment]
Oh boy. Yeah, I'm afraid my scripts aren't able to deal with genes breaking across scaffolds, because I've not dealt with such corner cases when writing the code... (err, I'm actually not sure whether multi-scaffold genes are possible in the GFF3 specifications.)

I've also noticed that some of your parents have a space between "1408" and "-RA", which is generally a bad idea, e.g. in these lines (aren't the spaces supposed to be "8"?)

```
Sc0000171 maker CDS 20794 20832 . + 0 ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
Sc0000171 maker CDS 22025 22047 . + 0 ID=Mcavernosa14088-RA:cds;Parent=Mcavernosa1408 -RA;
```

Honestly, over time, I've discovered that GFF3s from genome projects are terribly non-standard. Try your best in parsing information from the GFF3, and if dropping 300 weird genes allows the other 20,000 genes to be parsed properly, then I'd happily sacrifice them. My intuition tells me that dropping ~1-2% of the gene repertoire would have little effect on downstream analysis e.g. PCAs, hierarchical clustering, GO term analyses of differential methylated genes GO term analyses etc. If, however, 10% genes are affected, then I'd probably think of ways to parse the GFF3s better.

At this point I'll have to mention that my GFF3 parsing code works well only on the genomes I've worked on (because I'm able to troubleshoot problems and patch stuff up when things crash). I'd... be pleasantly surprised if my code worked on the M. cavernosa GFF3 post-removal of multi-scaffold genes. I'm afraid you'll have to put in some elbow grease (finger grease?) to get the code working... or try something else.

**Comment by jarcasariego** [comment]
The GFF3 has several issues that make it very hard to parse using your code. I'm working to try correct it but maybe I'll have to re-annotate the genome. I'm still figuring out what can I do before getting into the hurdling of redoing the annotation. 
Thank you very much for the help. I'll keep you informed of my progress. 
One other thing, I'm trying the SNP analysis on the pdae_dna_meth repository and there is one file that I don't know how to create. I was able to tabulate the genotypes after BIS-SNP analysis but was not able to find how can I create the BED file needed to extract the depth before filtering the SNPs. 
Cheers, 
Javier

**Comment by lyijin** [comment]
Oh indeed I wasn't clear in describing how the BED file was created. The actual command's lost to time; but I believe it was created with this command (let me know if it works for you so that I can update the README)

`sed 1d tabulated_genotypes.tsv | awk '{print $1"\t"$2-1"\t"$2}' > tabulated_genotypes.bed`

What this command is doing is that it is creating a faux-BED file from the tsv file. The tsv file has scaf at first column and pos at second, while BED wants scaf / (pos - 1) / pos as its 3 required columns (BED is 0-based in storing genomic positions).

This faux-BED file is then fed into samtools depth, which checks for coverage values at the specified positions.

... I think. Test it and see whether it works?

**Comment by jarcasariego** [comment]
It worked! Thanks. I generated all depth files for each sorted.bam. Next in the code you mention that these are merged to generate tabulated_depths.tsv. Is there a specific way to do this merge like with one of the tabulate_*.py scripts?

**Comment by lyijin** [comment]
Ah. I believe it'd be

`tabulate_tsvs.py (the_individual_depth_files) -k 0 1 -v 2 > tabulated_depths.tsv`

**Comment by jarcasariego** [comment]
Worked fine and was able to complete SNP analysis and filter by 5 criteria. Now I have to solve the gff3 issue before moving forward. Thanks a lot for the help. I'll keep in touch.
